observable/observable.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) The EfficientGo Authors.
+// Licensed under the Apache License 2.0.
+
+package e2eobs
+
+import (
+	"github.com/efficientgo/core/backoff"
+	"github.com/efficientgo/e2e"
+	e2emon "github.com/efficientgo/e2e/monitoring"
+	e2eprof "github.com/efficientgo/e2e/profiling"
+	"github.com/efficientgo/e2e/profiling/parcaconfig"
+)
+
+// Observable represents runnable that is both instrumented and profiled. Typically, all well-written Go services
+// are observable.
+type Observable struct {
+	e2e.Runnable
+	e2emon.Instrumented
+	e2eprof.Profiled
+}
+
+type opt struct {
+	instrOpts []e2emon.InstrumentedOption
+	profOpts  []e2eprof.ProfiledOption
+}
+
+// WithScheme allows adding customized scheme. "http" or "https" values allowed. "http" by default.
+// If "https" is specified, insecure TLS will be performed.
+func WithScheme(scheme string) Option {
+	return func(o *opt) {
+		o.instrOpts = append(o.instrOpts, e2emon.WithInstrumentedScheme(scheme))
+		o.profOpts = append(o.profOpts, e2eprof.WithProfiledScheme(scheme))
+	}
+}
+
+// WithMetricPath sets a custom path for metrics page. "/metrics" by default.
+func WithMetricPath(metricPath string) Option {
+	return func(o *opt) {
+		o.instrOpts = append(o.instrOpts, e2emon.WithInstrumentedMetricPath(metricPath))
+	}
+}
+
+// WithProfiledConfig sets a custom parca ProfilingConfig entry about this runnable. Empty by default (Parca defaults apply).
+func WithProfiledConfig(config parcaconfig.ProfilingConfig) Option {
+	return func(o *opt) {
+		o.profOpts = append(o.profOpts, e2eprof.WithProfiledConfig(config))
+	}
+}
+
+// WithInstrumentedWaitBackoff allows customizing wait backoff when accessing metric endpoint.
+func WithInstrumentedWaitBackoff(waitBackoff *backoff.Backoff) Option {
+	return func(o *opt) {
+		o.instrOpts = append(o.instrOpts, e2emon.WithInstrumentedWaitBackoff(waitBackoff))
+	}
+}
+
+type Option func(*opt)
+
+// AsObservable wraps e2e.Runnable with Observable.
+// If runnable is running during invocation AsObservable panics.
+// NOTE(bwplotka): Caller is expected to discard passed `r` runnable and use returned Observable.Runnable instead.
```

Comment:
Might be worth highlighting in README?

Reply:
Probably it's a programmatic nuance that usually does not matter for normal use. Especially if someone uses `e2eobs`